### ISSUE TITLE
compressed_pair doesn't work with an empty final struct

### ISSUE THREAD

**Issue by danieljames** [boostorg/utility#43]
As recently discussed on the user list, final classes break EBO. This test case causes a compile error:

    #include <boost/compressed_pair.hpp>

    struct final_type final {};

    int main() {
        boost::compressed_pair<final_type, int> x;
    }

This can be fixed for some compilers using a wrapper such as this one by Nevin Liber: https://github.com/nliber/cool/blob/master/ebo_wrapper.h using `boost::is_final` rather than `std::is_final`. Such a wrapper would also be useful for other libraries which implement EBO themselves.

`boost::is_final` from type traits supports recent clang and gcc, and will hopefully support recent Visual C++ 2013 and later with boostorg/type_traits#60. That will still leave some compilers where final is available but `boost::is_final` doesn't work. As Nevin suggested, the user can specialize `is_final` to support such types. I don't see any better solution for such compilers.

**Comment by danieljames** [comment]
Looking at the implementation, rather than using a wrapper class, it might be better to change the instantiation of `compressed_pair_switch` to check both `is_empty` and `is_final`. I think we can just use: `::boost::is_empty<T1>::value && !::boost::is_final<T1>::value`, I assume we don't have to worry about old compilers with poor support for integral constant expressions anymore.

**Comment by Lastique** [comment]
I would suggest testing for `BOOST_IS_FINAL` before using `boost::is_final`. Otherwise it will break `compressed_pair` for compilers that don't support final classes.

In fact, it might make sense to detect if `std::is_final` is available and only use it in `compressed_pair` when it is available. This way we're not depending on Boost.TypeTraits implementation details, which `BOOST_IS_FINAL` is.

**Comment by danieljames** [comment]
`boost::is_final` returns `false` when it isn't supported, so it will always do the right thing for compilers that don't support final classes.

Using `std::is_final` could be good, but we currently don't have a mechanism for detecting it, so it's unlikely we'll get better support than `boost::is_final` without a fair bit of work, e.g. libstdc++ doesn't have `std::is_final` in C++11 mode, but does have `boost::is_final`.

Utility already makes heavy use of Type Traits, so we won't be adding another dependency.

**Comment by Lastique** [comment]
> `boost::is_final` returns `false` when it isn't supported, so it will always do the right thing for compilers that don't support final classes.

Right, sorry, I didn't think it through enough.

**Comment by danieljames** [comment]
It looks like `boost::is_empty` doesn't work with final classes either.

**Comment by Lastique** [comment]
Looks like `boost::is_empty` is implemented as emulation even if a compiler intrinsic is available. @jzmaddock is there a reason for this?

**Comment by glenfe** [comment]
@danieljames We now also have `boost::empty_value` which is a slightly more flexible version of `ebo_helper`:
* https://www.boost.org/doc/libs/master/libs/core/doc/html/core/empty_value.html
* https://github.com/boostorg/core/blob/master/include/boost/core/empty_value.hpp

e.g. 1
```
template<class Allocator>
class type
    : empty_value<Allocator> {
```
e.g. 2
```
template<class InnerAllocator, class OuterAllocator, class OtherAllocator>
class type
    : empty_value<InnerAllocator, 1>
    , empty_value<OuterAllocator, 2>
    , empty_value<OtherAllocator, 3> {
```
That index is optional, and can be specified to create distinct base types so that the optimization still kicks in when all three Allocators are the same type.

**Comment by pdimov** [comment]
Wouldn't `compressed_pair` be easier as `template<class T1, class T2> class compressed_pair: empty_value<T1, 1>, empty_value<T2, 2>`?

**Comment by pdimov** [comment]
Also `template<class... T> compressed_tuple: empty_value<T, i> {}` with the appropriate `index_sequence` magic applied. :-)

**Comment by Lastique** [comment]
I think, the regular `boost::tuple` can implement compression, if it doesn't yet. `std::tuple` does.

**Comment by pdimov** [comment]
Some std:tuples do, some don't, I think?

boost::tuple is legacy C++03 and I don't think we can change it because I suspect much code relies on its internals. Although who knows.

**Comment by Lastique** [comment]
Even if someone depends on its implementation details, it shouldn't stop us from developing the library.

**Comment by glenfe** [comment]
> Wouldn't `compressed_pair` be easier as `template<class T1, class T2> class compressed_pair: empty_value<T1, 1>, empty_value<T2, 2>`?

Yes.

**Comment by Lastique** [comment]
I still think we should try and improve `boost::tuple`, even if it involves deprecating or changing some things. Tuples are supposed to be vocabulary types, and we already have a few implementations in different libraries. Let's not make another one.

**Comment by pdimov** [comment]
We have one in Hana. Tuples aren't exactly vocabulary types though. std::tuple is, the rest are tuple-likes.

Never mind. :-)

**Comment by glenfe** [comment]
I fixed it in #49 (https://github.com/boostorg/utility/pull/49). We can improve (simplify) compressed_pair, but that can happen later.
